### PR TITLE
PGD scheme

### PR DESCRIPTION
**Pull request by dlanzieri**
Add functions for the PGD implementation

### CODE REVIEW

**Comment by EiffL** [comment]
Thanks @dlanzieri for opening this PR :-) let me add a little bit of context. This PR aims to address issue #89 and provide the tools for PGD sharpening of flowpm simulations.

We still have a number of open issue:

- [ ] : compatibility with PGD parameters from madlens. Apparently we have different conventions somewhere because when applying the parameters obtained by biwei it doesn't give the expected result.

- [ ] : differentiability with respect to PGD parameters. We will probably want to optimize or even sample over PGD params but I think right now the PGD kernel is in simple python.

@modichirag do you have time for a review and try to check these two boxes? To merge this PR we want to at least have a set of PGD params that works as expected

**Comment by modichirag** [comment]
I tried fitting for the parameters in the last commit fc92237
The only thing we basically needed to do was make the kernel function tf expression, and so I moved it into tfpm instead of kernels (since kernels in kernels.py were simple np functions, but that can be moved back)

We need to figure out the loss function for fitting the parameters. 
https://arxiv.org/pdf/1804.00671.pdf
Appendix B in Biwei's paper suggests MSE of power spectrum, weighted by reference power spectrum. 
One place where we might differ is that our reference power spectrum comes from jax while for the paper it comes from higher resolution simulation and hence does not have cosmic variance. 
This matters on the largest scales, but it is possible that the response to PGD parameters is 0 due to suppressing large scales with the PGD kernel and so we might not need to worry about it.
We can also weigh it by 1 - k/knyquist to suppress smaller scales than we need i.e. k>2

This is what it seems to do for L=200, N=128, B=1, T=40 steps. 
image

This is what it seems to do for L=200, N=128, B=1, T=10 steps. 
image

I also asked Yu Feng. While MADlens uses PGD for post processing, c-fastpm uses it as a part of the simulation. We can try to do that if we want. To keep things simple, we can start with just fitting independent/different parameters at all redshifts before tying ourselves to some interpolating function. In his paper, Biwei gives parametric form in Eq. 4.1-4.5, but I am not sure how useful it is given that it is only for resolutions > 1.

**Comment by modichirag** [comment]
The last commit b52c8b9 adds online PGD in tfpm_pgd.py.
Online pgd means that after every drift step, we do a PGD update. Currently I have set it up so that it takes in the reference power spectra at all intermediate points and fits the PGD parameters independently at every redshift to match this reference spectra.
The results are sensible, so that is good. 

This is how it looks for 400 Mpc/h box, 10 steps, B=1, nc=128.
(My 200 Mpc/h runs did not match the spectra on the large scales even for normal N-Body, which is worrying...? This was what Denise was seeing for 50Mpc/h box)
image

It needs to be compared wether it's better to do online fitting or just fit each snapshot separately in post processing, as the previous commit was doing. I am hoping Denise can look into that?

Also, as Francois pointed out, we are comparing CIC-smoothed spectra with reference spectra.
Maybe we can smooth the reference spectra somehow?
Otherwise maybe we can run some test simulations on fastpm with high resolution and just save their snapshots on the low resolution of interest where we want to fit the parameters and use them to fit PGD parameters. This should not take a lot of effort and is probably easier thing in the short run.

**Comment by dlanzieri** [comment]
Thank you @modichirag for this review! I'll sure take a look at it. Just one thing, maybe I'm missing something,  if the last plot shows the ratio between our matter power spectrum and the reference power spectrum, what's the difference between the dashed and the solide lines?

File: flowpm/tfpm.py [hunk 1]
```diff
@@ -416,3 +590,96 @@ def nbody(cosmo,
...
+              nc,
+              box_size,
+              psref,
+              pgdparams,
+              pm_nc_factor=1,
+              niters=20,
+              return_intermediate_states=False,
+              name="NBody"):
+  """
+  Integrate the evolution of the state across the givent stages
```
Comment (EiffL) [inline flowpm/tfpm.py:607]: This lacks proper documentation of what this function does

File: scripts/fit_pgd_online.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+sys.path.append('../flowpm/')
+import flowpm
+import pickle
+from absl import app
+from absl import flags
+import flowpm.tfpower as tfpower
+import flowpm.scipy.interpolate as interpolate
+from flowpm.tfpower import linear_matter_power
+from flowpm import tfpm
+import jax_cosmo as jc
```
Comment (EiffL) [inline scripts/fit_pgd_online.py:16]: We shouldn't be using jaxcosmo, it's not a dependency in flowpm. Instead we have nbodykit if we want.

File: flowpm/tfpm.py [hunk 3]
```diff
@@ -416,3 +590,96 @@ def nbody(cosmo,
       return intermediate_states
     else:
       return state
+
+
+def nbody_pgd(cosmo,
```
Comment (EiffL) [inline flowpm/tfpm.py:595]: Ohhh wait.... what is this function supposed to be doing exactly? It fits and apply pgd online???

**Review by EiffL — COMMENTED** [review]
I have a few questions, I'll attempt a cleanup in a separate branch and make a PR to this one

File: flowpm/tfpm.py (hunk 3 above)
Comment (modichirag) [reply flowpm/tfpm.py:595]: Yeah, this is supposed to apply PGD after the drift step...so small correction to positions of the particles.
Currently it's fitting for these PGD parameters at every redshift. But you do not want to be the case when you are running a simulation since there you will ideally be feeding in the fitted parameters. 
So we need a separate script/file for fitting PGD parameters and applying PGD.
I would say apply_PGD should still be in tfpm.py but not fit_pgd.

Comment (EiffL) [reply flowpm/tfpm.py:595]: cool cool thanks :-)

**Comment by EiffL** [comment]
Ok I made some modifs to respond to my own comments above, and saved them there: #92

File: flowpm/tfpm.py [hunk 4]
```diff
@@ -7,7 +7,9 @@
 import tensorflow as tf
 from flowpm.tfbackground import f1, E, f2, Gf, gf, gf2, D1, D2, D1f
 from flowpm.utils import white_noise, c2r3d, r2c3d, cic_paint, cic_readout
-from flowpm.kernels import fftk, laplace_kernel, gradient_kernel, longrange_kernel
+from flowpm.kernels import fftk, laplace_kernel, gradient_kernel, longrange_kernel  #,  PGD_kernel
+
+from flowpm.pk import pk as pkl
```
[the remark this answers is not in the document]
Comment (dlanzieri) [reply flowpm/tfpm.py:12]: @EiffL The pgd function defined in tfpm.py returns the 3D power spectrum and uses the pkl to compute it

[the remark this answers is not in the document]
Comment (EiffL) [reply flowpm/tfpm.py:12]: Sure, :-) but the term "pkl" it doesn't mean anything to me here, so not a good name to give to a function. It could be misinterpreted as meaning linear pk

File: flowpm/tfpm.py [hunk 5]
```diff
@@ -231,14 +221,96 @@ def apply_longrange(x,
...
+  ks4 = ks**4
+  mask = (kk == 0).nonzero()
+  kk[mask] = 1
+  v = tf.exp(-kl2 / kk) * tf.exp(-kk**2 / ks4)
+  imask = (~(kk == 0)).astype(int)
+  v *= imask
+  return v
+
+
+def apply_PGD(x, delta_k, alpha, kl, ks, kvec=None, name="ApplyLongrange"):
```
[the remark this answers is not in the document]
Comment (dlanzieri) [reply flowpm/tfpm.py:256]: @EiffL It's a mistake from a copy/paste

File: scripts/fit_pgd_online.py (hunk 2 above)
Comment (dlanzieri) [reply scripts/fit_pgd_online.py:16]: can we compute the halofit power spectrum by nbodykit?

Comment (EiffL) [reply scripts/fit_pgd_online.py:16]: Yes, see my PR

File: flowpm/pk.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+def power_spectrum(field, kmin=5, dk=0.5, boxsize=False):
+  """
+    Calculate the powerspectra given real space field
+    
+    Args:
+        
+        field: real valued field 
+        kmin: minimum k-value for binned powerspectra
+        dk: differential in each kbin
+        shape: shape of field to calculate field (can be strangely shaped?)
```
Comment (EiffL) [inline flowpm/pk.py:48]:
```suggestion
```

File: flowpm/tfpm.py [hunk 7]
```diff
@@ -231,14 +219,64 @@ def apply_longrange(x,
...
+  alpha: float
+    Free parameter. Factor of proportionality between the displacement and the Particle-mesh force.
+
+  kl: float
+    Long range scale parameter
+    
+  ks: float
+    Short range scale parameter
+  """
+  # use the four point kernel to suppresse artificial growth of noise like terms
```
Comment (EiffL) [inline flowpm/tfpm.py:243]: what? is this comment relevant here?

File: flowpm/tfpm.py [hunk 8]
```diff
@@ -336,6 +368,64 @@ def force(cosmo,
     return state
 
 
+def PGD_correction(state,
+                   pgdparams,
+                   nc,
+                   pm_nc_factor=1,
+                   kvec=None,
+                   dtype=tf.float32,
+                   name="Force",
```
Comment (EiffL) [inline flowpm/tfpm.py:377]: This is not the correct name

Comment (EiffL) [reply flowpm/tfpm.py:377]:
```suggestion
                   name="PGD",
```

File: flowpm/tfpm.py (hunk 7 above)
Comment (EiffL) [reply flowpm/tfpm.py:243]:
```suggestion
```

**Review by EiffL — CHANGES_REQUESTED** [review]
Made a few minor suggestions, otherwise it looks good to go to me

File: scripts/fit_pgd.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    print("Fitted params (alpha, kl, ks)", params[-1])
+
+    plt.loglog(k, target_pk, "k")
+    plt.loglog(k, pk0, ':', label='starting')
+    plt.loglog(k, pk, '--', label='after n steps')
+    plt.grid(which='both')
+    plt.savefig('PGD_fit_%0.2f.png' % a)
+    plt.close()
+
+  pickle.dump(
```
Comment (modichirag) [inline scripts/fit_pgd.py:155]: Should we not be saving the boxsize and the grid size here as well?
Maybe also \omega_m and \sigma_8, but that depends on whether if we want to change cosmology/move to parametric form of the parameters at some point in the future.

Comment (EiffL) [reply scripts/fit_pgd.py:155]: yeah that's a good point, it really doesn't hurt to have more meta info

File: scripts/fit_pgd.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+def main(_):
+  cosmology = flowpm.cosmology.Planck15()
+  # Create a simple Planck15 cosmology without neutrinos, and makes sure sigma8
+  # is matched
+  nbdykit_cosmo = Cosmology.from_astropy(Planck15.clone(m_nu=0 * u.eV))
+  nbdykit_cosmo = nbdykit_cosmo.match(sigma8=cosmology.sigma8.numpy())
+
+  # Compute the k vectora that will be needed in the PGD fit
+  k, _ = flowpm.power_spectrum(
```
Comment (modichirag) [inline scripts/fit_pgd.py:87]: Minor issue that can totally be ignored and this has nothing to do with PGD, but maybe we want to use _initialize_pk or rename that function to get the parameters which are not dependent on the field. 
For instance, in addition to just extracting k-vector, there have been occassions when I needed to extract bin-edges or the number of modes in a bin.

File: scripts/fit_pgd.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+import flowpm.scipy.interpolate as interpolate
+
+from nbodykit.cosmology import Cosmology
+from nbodykit.cosmology.power.halofit import HalofitPower
+from astropy.cosmology import Planck15
+import astropy.units as u
+
+flags.DEFINE_string("filename", "results_fit_PGD.pkl", "Output filename")
+flags.DEFINE_integer("niter", 51, "Number of iterations of the first PGD fit")
+flags.DEFINE_integer("niter_refine", 11,
```
Comment (modichirag) [inline scripts/fit_pgd.py:21]: Seems a little severe to me and might not work out everytime but I guess we are leaving it to be fit and then gauge the results from the plot?

Comment (EiffL) [reply scripts/fit_pgd.py:21]: yeah... you mean the 11 iterations right? In my test, the parameters change very little from one snapshot to the next, but maybe it's because we have 40 steps

File: scripts/fit_pgd.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+  k, pk = flowpm.power_spectrum(
+      field,
+      boxsize=np.array([FLAGS.box_size] * 3),
+      kmin=np.pi / FLAGS.box_size,
+      dk=2 * np.pi / FLAGS.box_size)
+  # Averaging pk over realisations
+  pk = tf.reduce_mean(pk, axis=0)
+
+  # Step IV: compute loss
+  weight = 1 - k / (np.pi * FLAGS.nc / FLAGS.box_size)
```
Comment (modichirag) [inline scripts/fit_pgd.py:69]: In my experiments, I could not settle on the weight, unless you did. Note that this is not the weight that Biwei uses in his paper, it was something I came up with, adhoc. 
So we might want to keep this highlighted somehow so that we remember what is being done. Maybe use an argument (FLAGS) to the script instead of just keeping it under the wraps here.

Comment (EiffL) [reply scripts/fit_pgd.py:69]: I have spent exactly 0 minutes tuning the weights..... are you saying this is a really bad choice? why do we need a weight anyway actually?

Comment (modichirag) [reply scripts/fit_pgd.py:69]: I am not saying this is a bad choice. I think this one worked for me. But this is adhoc and we should be aware that it is there.

The reason I started including weights is because there are competing effects that we do not want to correct with PGD - on large scales, your PS will not agree with theory spectra due to cosmic variance; on small scales, we have the CIC window kicking in for which we are not correctly compensating  (are we?). weights were a way to downweight the loss function on these scales which are dominated by these effects so that PGD can focus on what we actually care about. 
Incidentally...fitting for a high resolution simulation instead of halofit will automatically take care of these effects ;)

File: scripts/fit_pgd.py (hunk 10 above)
Comment (EiffL) [reply scripts/fit_pgd.py:87]: sure.... but if not needed for this script, I would opt for simplicity. the _initialize_pk is meant to remain secret.

**Review by dlanzieri — COMMENTED** [review]
it looks good to go to me,  we should proceed

File: scripts/fit_pgd.py (hunk 9 above)
Comment (EiffL) [reply scripts/fit_pgd.py:155]: we are adding cosmological parameters as well as size and resolution of the simulation

File: scripts/fit_pgd.py (hunk 12 above)
Comment (EiffL) [reply scripts/fit_pgd.py:69]: ok. well... so this is not great, and we should really use the cic painting compensation on small scales. On large scales, in the current script we are averaging the power spectrum over several realisations, so that should take care of cosmic variance more or less.

For now, I'm following your suggestion, I've made your custom weight a parameter. We will open another issue to improve this further.

**Review by EiffL — APPROVED** [review]
Thank you Denise for looking at this. @modichirag we addressed some of your concerns, and I'm flagging the issue of proper weighting to be separate issue

**Review by dlanzieri — COMMENTED** [review]
@EiffL Thank you for these last changes :)
